{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c341e94",
   "metadata": {},
   "source": [
    "# Chaining Search Terms\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8387f59",
   "metadata": {},
   "source": [
    "Before we do any work, we need to import several functions from cdapython:\n",
    "\n",
    "\n",
    "- `Q` and `query` which power the search\n",
    "- `columns` which lets us view entity field names\n",
    "- `unique_terms` which lets view entity field contents\n",
    "    \n",
    "We're also asking cdapython to report it's version so we can be sure we're using the one we mean to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28a10e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.5.9\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "print(cdapython.__version__)\n",
    "Q.set_host_url(\"http://35.192.60.10:8080/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0810c61c",
   "metadata": {},
   "source": [
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "\n",
    "- **[Q.run()](../../../Documentation/usage/#qrun)** returns data for the specified search \n",
    "- **[Q.count()](../../../Documentation/usage/#qcounts)** returns summary information (counts) data that fit the specified search\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8084b3",
   "metadata": {},
   "source": [
    "## Chaining\n",
    "\n",
    "We're going to build on our [previous search](../BasicSearch) to see what information exists about cancers that were first diagnosed in the brain. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d804f764",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "736ad485",
   "metadata": {},
   "source": [
    "Previously we looked at subject, research_subject, specimen and file results seperately, but we can also combine these. \n",
    "\n",
    "Let's say what we're really interested in is finding analysis done on specimens, so we're looking for files that belong to specimens that match our search. If we search the files endpoint directly, we'll get back all files that meet our search criteria, regardless of whether they are specimens specficially. Instead, we can chain our query to the specimen endpoint and then to the files endpoint and get the combined result:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7284df72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3622 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 3845d4c4-0e99-490c-b2f3-fa59288df875\n",
       "            Query:SELECT results.* EXCEPT(rn) FROM (SELECT ROW_NUMBER() OVER (PARTITION BY all_Files_v3_0_w_RS.id) as rn, all_Files_v3_0_w_RS.id AS id, all_Files_v3_0_w_RS.identifier AS identifier, all_Files_v3_0_w_RS.label AS label, all_Files_v3_0_w_RS.data_category AS data_category, all_Files_v3_0_w_RS.data_type AS data_type, all_Files_v3_0_w_RS.file_format AS file_format, all_Files_v3_0_w_RS.associated_project AS associated_project, all_Files_v3_0_w_RS.drs_uri AS drs_uri, all_Files_v3_0_w_RS.byte_size AS byte_size, all_Files_v3_0_w_RS.checksum AS checksum, all_Files_v3_0_w_RS.data_modality AS data_modality, all_Files_v3_0_w_RS.imaging_modality AS imaging_modality, all_Files_v3_0_w_RS.dbgap_accession_number AS dbgap_accession_number FROM gdc-bq-sample.dev.all_Subjects_v3_0_w_RS AS all_Subjects_v3_0_w_RS LEFT JOIN UNNEST(all_Subjects_v3_0_w_RS.ResearchSubject) AS _ResearchSubject LEFT JOIN UNNEST(_ResearchSubject.Specimen) AS _ResearchSubject_Specimen LEFT JOIN UNNEST(_ResearchSubject_Specimen.Files) AS _ResearchSubject_Specimen_Files  LEFT JOIN gdc-bq-sample.dev.all_Files_v3_0_w_RS AS all_Files_v3_0_w_RS ON all_Files_v3_0_w_RS.id = _ResearchSubject_Specimen_Files WHERE (UPPER(_ResearchSubject.primary_diagnosis_site) = UPPER('brain'))) as results WHERE rn = 1\n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 50495\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles =  myquery.specimen.files.run()\n",
    "myqueryspecimenfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d9154b",
   "metadata": {},
   "source": [
    "We get back 50495 files that belong to specimens that meet our search criteria. As before, we can preview the results by using the `.to_dataframe()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "410b0e2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1676bc16-a120-4951-bcb9-b3e0599b45af</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1676bc16-a120-495...</td>\n",
       "      <td>7dee3b3f-7f68-493b-a99c-49ae65675609.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:1676bc16-a120-4951-bcb9-b3e0599b...</td>\n",
       "      <td>1051</td>\n",
       "      <td>ecabcaa16c4f79262c996b5df904d3c3</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1825d954-56e3-4faa-991e-56e28d23f31e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1825d954-56e3-4fa...</td>\n",
       "      <td>415b62cf-1ab3-4432-9fd4-c326718d8142.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:1825d954-56e3-4faa-991e-56e28d23...</td>\n",
       "      <td>22932</td>\n",
       "      <td>78fb97b022c9790c89888ba62bfb222e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3c4c6124-e58f-497a-8a20-e5ade9caa765</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3c4c6124-e58f-497...</td>\n",
       "      <td>c7a985b2-bc10-4e58-86e8-0ef41f41f80c.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>drs://dg.4DFC:3c4c6124-e58f-497a-8a20-e5ade9ca...</td>\n",
       "      <td>1053</td>\n",
       "      <td>d7dabf1852034e3fc333ae68b38f2281</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9148502d-cb91-41c6-b382-62639010ff4e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '9148502d-cb91-41c...</td>\n",
       "      <td>e1c60464-3f4f-400b-bbfa-0e96f826d901.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:9148502d-cb91-41c6-b382-62639010...</td>\n",
       "      <td>3825</td>\n",
       "      <td>20d837abab0ffa04e06e5de5f35108e8</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7aef2592-bb88-44fe-886c-22e0e3fd4aac</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7aef2592-bb88-44f...</td>\n",
       "      <td>9010b2f6-5eb0-4c68-b42a-fb271d8f5ba2.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>drs://dg.4DFC:7aef2592-bb88-44fe-886c-22e0e3fd...</td>\n",
       "      <td>1051</td>\n",
       "      <td>3f36395c16a0fc18171d490f1039efcb</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>d91dbc00-747a-4c78-9fe9-d3d599df2656</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd91dbc00-747a-4c7...</td>\n",
       "      <td>41a6475c-09bf-4b0c-b568-a5ce5d8f7a23.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-GRCC</td>\n",
       "      <td>drs://dg.4DFC:d91dbc00-747a-4c78-9fe9-d3d599df...</td>\n",
       "      <td>3465</td>\n",
       "      <td>fa2269804bffbba348da639311efe254</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>87b76b2b-7f27-49b1-8989-c1c3820e5661</td>\n",
       "      <td>[{'system': 'GDC', 'value': '87b76b2b-7f27-49b...</td>\n",
       "      <td>39444fc1-0c79-44c6-a1dc-9f86e72835e1.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-GRCC</td>\n",
       "      <td>drs://dg.4DFC:87b76b2b-7f27-49b1-8989-c1c3820e...</td>\n",
       "      <td>2196</td>\n",
       "      <td>68f202798873dfdd85b5680cb6263388</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>c609be50-3cc3-447a-acbc-4d82b4c30e72</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'c609be50-3cc3-447...</td>\n",
       "      <td>fa186f29-1bd9-4b82-9daf-d606a6ddef11.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-GRCC</td>\n",
       "      <td>drs://dg.4DFC:c609be50-3cc3-447a-acbc-4d82b4c3...</td>\n",
       "      <td>3565</td>\n",
       "      <td>b1b7e7322180b6764e27faac63e77bc2</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>e47498b5-27ae-46f7-8ee5-cf310360c598</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e47498b5-27ae-46f...</td>\n",
       "      <td>9a9009af-e4ce-4477-9af7-851c82e61802.genie.ali...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Annotated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>GENIE-MDA</td>\n",
       "      <td>drs://dg.4DFC:e47498b5-27ae-46f7-8ee5-cf310360...</td>\n",
       "      <td>1052</td>\n",
       "      <td>919e44b2808e2de2281ba9275698ba1d</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>8a399f7b-ee78-4735-97d4-70f4ad28fe68</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8a399f7b-ee78-473...</td>\n",
       "      <td>7995a73b-b4e4-419b-badb-668a33e48f76.genie.ali...</td>\n",
       "      <td>Copy Number Variation</td>\n",
       "      <td>Gene Level Copy Number Scores</td>\n",
       "      <td>TSV</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>drs://dg.4DFC:8a399f7b-ee78-4735-97d4-70f4ad28...</td>\n",
       "      <td>22909</td>\n",
       "      <td>b1e1dd883334713bc445dc969922f70f</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1676bc16-a120-4951-bcb9-b3e0599b45af   \n",
       "1   1825d954-56e3-4faa-991e-56e28d23f31e   \n",
       "2   3c4c6124-e58f-497a-8a20-e5ade9caa765   \n",
       "3   9148502d-cb91-41c6-b382-62639010ff4e   \n",
       "4   7aef2592-bb88-44fe-886c-22e0e3fd4aac   \n",
       "..                                   ...   \n",
       "95  d91dbc00-747a-4c78-9fe9-d3d599df2656   \n",
       "96  87b76b2b-7f27-49b1-8989-c1c3820e5661   \n",
       "97  c609be50-3cc3-447a-acbc-4d82b4c30e72   \n",
       "98  e47498b5-27ae-46f7-8ee5-cf310360c598   \n",
       "99  8a399f7b-ee78-4735-97d4-70f4ad28fe68   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '1676bc16-a120-495...   \n",
       "1   [{'system': 'GDC', 'value': '1825d954-56e3-4fa...   \n",
       "2   [{'system': 'GDC', 'value': '3c4c6124-e58f-497...   \n",
       "3   [{'system': 'GDC', 'value': '9148502d-cb91-41c...   \n",
       "4   [{'system': 'GDC', 'value': '7aef2592-bb88-44f...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': 'd91dbc00-747a-4c7...   \n",
       "96  [{'system': 'GDC', 'value': '87b76b2b-7f27-49b...   \n",
       "97  [{'system': 'GDC', 'value': 'c609be50-3cc3-447...   \n",
       "98  [{'system': 'GDC', 'value': 'e47498b5-27ae-46f...   \n",
       "99  [{'system': 'GDC', 'value': '8a399f7b-ee78-473...   \n",
       "\n",
       "                                                label  \\\n",
       "0   7dee3b3f-7f68-493b-a99c-49ae65675609.genie.ali...   \n",
       "1   415b62cf-1ab3-4432-9fd4-c326718d8142.genie.ali...   \n",
       "2   c7a985b2-bc10-4e58-86e8-0ef41f41f80c.genie.ali...   \n",
       "3   e1c60464-3f4f-400b-bbfa-0e96f826d901.genie.ali...   \n",
       "4   9010b2f6-5eb0-4c68-b42a-fb271d8f5ba2.genie.ali...   \n",
       "..                                                ...   \n",
       "95  41a6475c-09bf-4b0c-b568-a5ce5d8f7a23.genie.ali...   \n",
       "96  39444fc1-0c79-44c6-a1dc-9f86e72835e1.genie.ali...   \n",
       "97  fa186f29-1bd9-4b82-9daf-d606a6ddef11.genie.ali...   \n",
       "98  9a9009af-e4ce-4477-9af7-851c82e61802.genie.ali...   \n",
       "99  7995a73b-b4e4-419b-badb-668a33e48f76.genie.ali...   \n",
       "\n",
       "                  data_category                          data_type  \\\n",
       "0   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "1         Copy Number Variation      Gene Level Copy Number Scores   \n",
       "2   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "3   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "4   Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "..                          ...                                ...   \n",
       "95  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "96  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "97  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "98  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   \n",
       "99        Copy Number Variation      Gene Level Copy Number Scores   \n",
       "\n",
       "   file_format associated_project  \\\n",
       "0          MAF          GENIE-MSK   \n",
       "1          TSV          GENIE-MSK   \n",
       "2          MAF          GENIE-JHU   \n",
       "3          MAF          GENIE-MSK   \n",
       "4          MAF          GENIE-MSK   \n",
       "..         ...                ...   \n",
       "95         MAF         GENIE-GRCC   \n",
       "96         MAF         GENIE-GRCC   \n",
       "97         MAF         GENIE-GRCC   \n",
       "98         MAF          GENIE-MDA   \n",
       "99         TSV         GENIE-DFCI   \n",
       "\n",
       "                                              drs_uri byte_size  \\\n",
       "0   drs://dg.4DFC:1676bc16-a120-4951-bcb9-b3e0599b...      1051   \n",
       "1   drs://dg.4DFC:1825d954-56e3-4faa-991e-56e28d23...     22932   \n",
       "2   drs://dg.4DFC:3c4c6124-e58f-497a-8a20-e5ade9ca...      1053   \n",
       "3   drs://dg.4DFC:9148502d-cb91-41c6-b382-62639010...      3825   \n",
       "4   drs://dg.4DFC:7aef2592-bb88-44fe-886c-22e0e3fd...      1051   \n",
       "..                                                ...       ...   \n",
       "95  drs://dg.4DFC:d91dbc00-747a-4c78-9fe9-d3d599df...      3465   \n",
       "96  drs://dg.4DFC:87b76b2b-7f27-49b1-8989-c1c3820e...      2196   \n",
       "97  drs://dg.4DFC:c609be50-3cc3-447a-acbc-4d82b4c3...      3565   \n",
       "98  drs://dg.4DFC:e47498b5-27ae-46f7-8ee5-cf310360...      1052   \n",
       "99  drs://dg.4DFC:8a399f7b-ee78-4735-97d4-70f4ad28...     22909   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   ecabcaa16c4f79262c996b5df904d3c3       Genomic             None   \n",
       "1   78fb97b022c9790c89888ba62bfb222e       Genomic             None   \n",
       "2   d7dabf1852034e3fc333ae68b38f2281       Genomic             None   \n",
       "3   20d837abab0ffa04e06e5de5f35108e8       Genomic             None   \n",
       "4   3f36395c16a0fc18171d490f1039efcb       Genomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  fa2269804bffbba348da639311efe254       Genomic             None   \n",
       "96  68f202798873dfdd85b5680cb6263388       Genomic             None   \n",
       "97  b1b7e7322180b6764e27faac63e77bc2       Genomic             None   \n",
       "98  919e44b2808e2de2281ba9275698ba1d       Genomic             None   \n",
       "99  b1e1dd883334713bc445dc969922f70f       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number  \n",
       "0                    None  \n",
       "1                    None  \n",
       "2                    None  \n",
       "3                    None  \n",
       "4                    None  \n",
       "..                    ...  \n",
       "95                   None  \n",
       "96                   None  \n",
       "97                   None  \n",
       "98                   None  \n",
       "99                   None  \n",
       "\n",
       "[100 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myqueryspecimenfiles.to_dataframe()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
