{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f90b531",
   "metadata": {},
   "source": [
    "Before we do any work, we need to import several functions from cdapython:\n",
    "- `Q` and `query` which power the search\n",
    "- `columns` which lets us view entity field names\n",
    "- `unique_terms` which lets view entity field contents\n",
    "\n",
    "We're also asking cdapython to report it's version so we can be sure we're using the one we mean to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5245b7b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/amanda.charbonneau/github/cda-python/cdapython/__init__.py\n",
      "2022.5.23\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms,query, constantVariables\n",
    "import cdapython\n",
    "print(cdapython.__file__)\n",
    "print(cdapython.__version__)\n",
    "integration_host = \"http://35.192.60.10:8080/\"\n",
    "Q.set_host_url(integration_host)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d538da92",
   "metadata": {},
   "source": [
    "CDA data comes from three sources:\n",
    "- The [Proteomic Data Commons](https://proteomic.datacommons.cancer.gov/pdc/) (PDC)\n",
    "- The [Genomic Data Commons](https://gdc.cancer.gov/) (GDC)\n",
    "- The [Imaging Data Commons](https://datacommons.cancer.gov/repository/imaging-data-commons) (IDC)\n",
    "\n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "- `subject`: Specific, unique, individuals\n",
    "- `researchsubject`: Study-individual aggregate entities. A Subject who was part of three studies will appear as three ResearchSubjects\n",
    "- `specimen`: Samples taken from individual\n",
    "- `file`: Data about Subjects, ResearchSubjects, Specimens, and their associated information\n",
    "\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "\n",
    "- `diagnosis`: Information about what medical diagnosis a researchsubject has\n",
    "- `treatment`: Information about what medical treatment(s) were performed for a given diagnosis\n",
    "\n",
    "If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specfic format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.\n",
    "\n",
    "In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.\n",
    "\n",
    "In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76026a3",
   "metadata": {},
   "source": [
    "## Basic search with endpoints\n",
    "\n",
    "Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c92a98ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e855573",
   "metadata": {
    "tags": [
     "Help!"
    ]
   },
   "source": [
    "\n",
    "<div style=\"background-color:#6ce6b9;color:black;padding:20px;\">\n",
    "<h3>Where did those terms come from?</h3>\n",
    "    \n",
    "If you aren't sure how we knew what terms to put in our search, please refer back to the <a href=\"../SearchTerms\">What search terms are available?</a> notebook. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d483e944",
   "metadata": {},
   "source": [
    "### subject\n",
    "Now we can use that query to search any of information types. Let's start by looking at what Subjects meet our criteria. To do that, we will send our query to the subject endpoint, then ask for it to run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6d9137aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3748 ms\n"
     ]
    }
   ],
   "source": [
    "subjectresults = myquery.subject.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf7fb60",
   "metadata": {},
   "source": [
    "We saved the output in a variable `subjectresults`, so we don't get much visible output. To see what our results are, we need to look into the variable. The simplest way is to call `subjectresults` directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cff572d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 6b1f5113-18d4-40f0-9a49-4062a5653fea\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 2314\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5186036c",
   "metadata": {},
   "source": [
    "This output tells us our QueryID, which we don't really need, but the computer does to track our questions. Then it tells us four parameters that describe our results:\n",
    "\n",
    "---\n",
    "\n",
    "- **Offset:** This is how many rows of information we've told the query to skip in the data, here we didn't tell it to skip anything, so the offset is zero\n",
    "- **Count:** This is how many rows the current page of our results table has. To keep searches fast, we default to pages with 100 rows.\n",
    "- **Total Row Count:** This is how many rows are in the full results table\n",
    "- **More pages:** This is always a True or False. False means that our current page has all the available results. True means that we will see only the first 100 results in this table, and will need to page through for more.\n",
    "\n",
    "---\n",
    "    \n",
    "Now that we've seen the metadata about our results, let's look at the actual table. The easiest way to do this is by using the python function `.to_dataframe()` on our `subjectresults` variable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e97e19f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACRIN-FMISO-Brain-010</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'ACRIN-FMISO-Brain...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[acrin_fmiso_brain]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C140343</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'C140343'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>None</td>\n",
       "      <td>[Proteogenomic Analysis of Pediatric Brain Can...</td>\n",
       "      <td>Alive</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-01327</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'C3L-01327'}, {'sy...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27160</td>\n",
       "      <td>[cptac_gbm, CPTAC3-Discovery, CPTAC-3]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>646</td>\n",
       "      <td>Cancer Related</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C808110</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'C808110'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>None</td>\n",
       "      <td>[Proteogenomic Analysis of Pediatric Brain Can...</td>\n",
       "      <td>Alive</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-GRCC-e6a259ca</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-GRCC-e6a259...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not allowed to collect</td>\n",
       "      <td>not allowed to collect</td>\n",
       "      <td>-17166</td>\n",
       "      <td>[GENIE-GRCC]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>QIN-GBM-TR-01</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'QIN-GBM-TR-01'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[qin_gbm_treatment_response]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>RIDER Neuro MRI-2579032163</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'RIDER Neuro MRI-2...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[rider_neuro_mri]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-02-0336</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-02-0336'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[TCGA-GBM]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-02-0446</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-02-0446'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-22541</td>\n",
       "      <td>[TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>282</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-06-0397</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-0397'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20845</td>\n",
       "      <td>[TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>274</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            id  \\\n",
       "0        ACRIN-FMISO-Brain-010   \n",
       "1                      C140343   \n",
       "2                    C3L-01327   \n",
       "3                      C808110   \n",
       "4          GENIE-GRCC-e6a259ca   \n",
       "..                         ...   \n",
       "95               QIN-GBM-TR-01   \n",
       "96  RIDER Neuro MRI-2579032163   \n",
       "97                TCGA-02-0336   \n",
       "98                TCGA-02-0446   \n",
       "99                TCGA-06-0397   \n",
       "\n",
       "                                           identifier       species     sex  \\\n",
       "0   [{'system': 'IDC', 'value': 'ACRIN-FMISO-Brain...  Homo sapiens    None   \n",
       "1             [{'system': 'PDC', 'value': 'C140343'}]  Homo sapiens    male   \n",
       "2   [{'system': 'GDC', 'value': 'C3L-01327'}, {'sy...  Homo sapiens    male   \n",
       "3             [{'system': 'PDC', 'value': 'C808110'}]  Homo sapiens  female   \n",
       "4   [{'system': 'GDC', 'value': 'GENIE-GRCC-e6a259...  Homo sapiens  female   \n",
       "..                                                ...           ...     ...   \n",
       "95      [{'system': 'IDC', 'value': 'QIN-GBM-TR-01'}]  Homo sapiens    None   \n",
       "96  [{'system': 'IDC', 'value': 'RIDER Neuro MRI-2...  Homo sapiens    None   \n",
       "97       [{'system': 'GDC', 'value': 'TCGA-02-0336'}]  Homo sapiens    None   \n",
       "98       [{'system': 'GDC', 'value': 'TCGA-02-0446'}]  Homo sapiens    male   \n",
       "99       [{'system': 'GDC', 'value': 'TCGA-06-0397'}]  Homo sapiens  female   \n",
       "\n",
       "                      race               ethnicity days_to_birth  \\\n",
       "0                     None                    None          None   \n",
       "1                    white  not hispanic or latino          None   \n",
       "2                    white  not hispanic or latino        -27160   \n",
       "3                    white  not hispanic or latino          None   \n",
       "4   not allowed to collect  not allowed to collect        -17166   \n",
       "..                     ...                     ...           ...   \n",
       "95                    None                    None          None   \n",
       "96                    None                    None          None   \n",
       "97                    None                    None          None   \n",
       "98                   white  not hispanic or latino        -22541   \n",
       "99                   white  not hispanic or latino        -20845   \n",
       "\n",
       "                           subject_associated_project  vital_status  \\\n",
       "0                                 [acrin_fmiso_brain]          None   \n",
       "1   [Proteogenomic Analysis of Pediatric Brain Can...         Alive   \n",
       "2              [cptac_gbm, CPTAC3-Discovery, CPTAC-3]          Dead   \n",
       "3   [Proteogenomic Analysis of Pediatric Brain Can...         Alive   \n",
       "4                                        [GENIE-GRCC]  Not Reported   \n",
       "..                                                ...           ...   \n",
       "95                       [qin_gbm_treatment_response]          None   \n",
       "96                                  [rider_neuro_mri]          None   \n",
       "97                                         [TCGA-GBM]          None   \n",
       "98                                         [TCGA-GBM]          Dead   \n",
       "99                                         [TCGA-GBM]          Dead   \n",
       "\n",
       "   age_at_death  cause_of_death  \n",
       "0          None            None  \n",
       "1          None    Not Reported  \n",
       "2           646  Cancer Related  \n",
       "3          None    Not Reported  \n",
       "4          None            None  \n",
       "..          ...             ...  \n",
       "95         None            None  \n",
       "96         None            None  \n",
       "97         None            None  \n",
       "98          282            None  \n",
       "99          274            None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44a65df",
   "metadata": {},
   "source": [
    "By default `to_dataframe()` shows us the first and last five rows for the first page of our results, so we can easily preview our data.\n",
    "\n",
    "Since we queried the Subject endpoint, our default results tell us Subject level information, that is, information about unique individuals: their sex, race, age, species, etc. The `id` column tells us the unique identifier for each individual. The identifier column has nested information about what study or studies a Subject participated in, and will list all of their researchsubject identifiers. \n",
    "\n",
    "The `to_dataframe()` function converts the results to a pandas dataframe. So if we save the dataframe to a variable, we can use any pandas functions to explore it. For example, lets see whether any of the Subjects in our first 100 results are black or african american. First we'll save the results to a dataframe, then subset that dataframe to only show rows where the word \"black\" appears in the \"race\" column. \"NAs\" which are shown as \"None\" in these tables, so for our filter to work, we'll need to specifically tell it to ignore NAs. We're also telling it we want the word \"black\" regardless of capitalization with `case=False`:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4975c1e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>age_at_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>TCGA-06-A7TL</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-A7TL'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-11297</td>\n",
       "      <td>[TCGA-GBM]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>TCGA-26-A7UX</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-26-A7UX'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17365</td>\n",
       "      <td>[TCGA-GBM]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                    identifier       species  \\\n",
       "38  TCGA-06-A7TL  [{'system': 'GDC', 'value': 'TCGA-06-A7TL'}]  Homo sapiens   \n",
       "69  TCGA-26-A7UX  [{'system': 'GDC', 'value': 'TCGA-26-A7UX'}]  Homo sapiens   \n",
       "\n",
       "       sex                       race               ethnicity days_to_birth  \\\n",
       "38  female  black or african american  not hispanic or latino        -11297   \n",
       "69    male  black or african american  not hispanic or latino        -17365   \n",
       "\n",
       "   subject_associated_project vital_status age_at_death cause_of_death  \n",
       "38                 [TCGA-GBM]        Alive         None           None  \n",
       "69                 [TCGA-GBM]        Alive         None           None  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectdata = subjectresults.to_dataframe()\n",
    "subjectdata[subjectdata['race'].str.contains(\"black\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8750617",
   "metadata": {},
   "source": [
    "There are three subjects in our first hundred results that meet the criteria. If we just want to be sure that the data contains some value, this might be good enough. But often we want to search the entire set of results and not just the first page. \n",
    "\n",
    "We'll cover how to work with large results dataframes in the [Pagination](../Pagination.ipynb) notebook. Or, learn how to get summary information from search results in the [Data Summaries](../DataSummaries.ipynb) notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05e52f3f",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A subject is a specific, unique individual: for e.g. a single human. When consent allows, a given entity will have a single subject ID that can be connected to all their studies and data across all datasets</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this subject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the subject had there</li>\n",
    "  <li><b>species:</b> The species of the subject</li>\n",
    "  <li><b>sex:</b> A reference to the biological sex of the donor organism. </li>\n",
    "  <li><b>race:</b> The race of the subject</li>\n",
    "  <li><b>ethnicity:</b> The ethnicity of the subject</li>\n",
    "  <li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days</li>\n",
    "  <li><b>subject_associated_project:</b> An embedded array of the names of projects (studies) the subject was part of</li>\n",
    "  <li><b>vital_status:</b> Whether the subject is alive</li>\n",
    "  <li><b>age_at_death:</b> The number of days after first enrollment that the subject died</li>\n",
    "  <li><b>cause_of_death:</b> The cause of death, if known</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73745aa6",
   "metadata": {},
   "source": [
    "### researchsubject\n",
    "\n",
    "If we're interested in what researchsubjects meet our critera, we can also run our query against the researchsubject endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a2d8c874",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3490 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 35ba8bad-9d0d-4fb5-9b9d-c2445caf707d\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 2923\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults = myquery.researchsubject.run()\n",
    "researchsubjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "977f36fe",
   "metadata": {},
   "source": [
    "Now we see that our 2314 subjects have 2923 researchsubjects between them, that means that some, but not all, of our subjects were participants in more than one study. Let's peek at the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a20f7e62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>104c3a87-2139-11ea-aee1-0e1aae319e49</td>\n",
       "      <td>[{'system': 'PDC', 'value': '104c3a87-2139-11e...</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-03407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>153c2442-ea61-4b5e-8c5c-c71d287c6055</td>\n",
       "      <td>[{'system': 'GDC', 'value': '153c2442-ea61-4b5...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-28-5209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1edfe67c-1142-4498-927c-9b7b426af7a3</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1edfe67c-1142-449...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-02-0339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3d58ded1-b7cf-420c-ba59-5b424683f0df</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3d58ded1-b7cf-420...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-HT-7686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>559a8f6f-dd60-43b7-ae09-16cc4206064c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '559a8f6f-dd60-43b...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-HT-A74L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>723cad09-aa2d-4492-b19d-661b310e1c58</td>\n",
       "      <td>[{'system': 'GDC', 'value': '723cad09-aa2d-449...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-HT-A74O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>7338e476-10ff-444a-a16f-429d28355f65</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7338e476-10ff-444...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-28-5213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>834e560a-a506-456d-baf2-a206ca058153</td>\n",
       "      <td>[{'system': 'GDC', 'value': '834e560a-a506-456...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>TCGA-DU-7010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ACRIN-DSC-MR-Brain-037__acrin_dsc_mr_brain</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'ACRIN-DSC-MR-Brai...</td>\n",
       "      <td>acrin_dsc_mr_brain</td>\n",
       "      <td>None</td>\n",
       "      <td>Brain</td>\n",
       "      <td>ACRIN-DSC-MR-Brain-037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>C3L-01327__cptac_gbm</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-01327'}]</td>\n",
       "      <td>cptac_gbm</td>\n",
       "      <td>None</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-01327</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            id  \\\n",
       "0         104c3a87-2139-11ea-aee1-0e1aae319e49   \n",
       "1         153c2442-ea61-4b5e-8c5c-c71d287c6055   \n",
       "2         1edfe67c-1142-4498-927c-9b7b426af7a3   \n",
       "3         3d58ded1-b7cf-420c-ba59-5b424683f0df   \n",
       "4         559a8f6f-dd60-43b7-ae09-16cc4206064c   \n",
       "..                                         ...   \n",
       "95        723cad09-aa2d-4492-b19d-661b310e1c58   \n",
       "96        7338e476-10ff-444a-a16f-429d28355f65   \n",
       "97        834e560a-a506-456d-baf2-a206ca058153   \n",
       "98  ACRIN-DSC-MR-Brain-037__acrin_dsc_mr_brain   \n",
       "99                        C3L-01327__cptac_gbm   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'PDC', 'value': '104c3a87-2139-11e...   \n",
       "1   [{'system': 'GDC', 'value': '153c2442-ea61-4b5...   \n",
       "2   [{'system': 'GDC', 'value': '1edfe67c-1142-449...   \n",
       "3   [{'system': 'GDC', 'value': '3d58ded1-b7cf-420...   \n",
       "4   [{'system': 'GDC', 'value': '559a8f6f-dd60-43b...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '723cad09-aa2d-449...   \n",
       "96  [{'system': 'GDC', 'value': '7338e476-10ff-444...   \n",
       "97  [{'system': 'GDC', 'value': '834e560a-a506-456...   \n",
       "98  [{'system': 'IDC', 'value': 'ACRIN-DSC-MR-Brai...   \n",
       "99          [{'system': 'IDC', 'value': 'C3L-01327'}]   \n",
       "\n",
       "   member_of_research_project primary_diagnosis_condition  \\\n",
       "0            CPTAC3-Discovery                Glioblastoma   \n",
       "1                    TCGA-GBM                     Gliomas   \n",
       "2                    TCGA-GBM                     Gliomas   \n",
       "3                    TCGA-LGG                     Gliomas   \n",
       "4                    TCGA-LGG                     Gliomas   \n",
       "..                        ...                         ...   \n",
       "95                   TCGA-LGG                     Gliomas   \n",
       "96                   TCGA-GBM                     Gliomas   \n",
       "97                   TCGA-LGG                     Gliomas   \n",
       "98         acrin_dsc_mr_brain                        None   \n",
       "99                  cptac_gbm                        None   \n",
       "\n",
       "   primary_diagnosis_site              subject_id  \n",
       "0                   Brain               C3L-03407  \n",
       "1                   Brain            TCGA-28-5209  \n",
       "2                   Brain            TCGA-02-0339  \n",
       "3                   Brain            TCGA-HT-7686  \n",
       "4                   Brain            TCGA-HT-A74L  \n",
       "..                    ...                     ...  \n",
       "95                  Brain            TCGA-HT-A74O  \n",
       "96                  Brain            TCGA-28-5213  \n",
       "97                  Brain            TCGA-DU-7010  \n",
       "98                  Brain  ACRIN-DSC-MR-Brain-037  \n",
       "99                  Brain               C3L-01327  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4564beee",
   "metadata": {},
   "source": [
    "Each row from the researchsubject endpoint results tells us about a subject in a given study. Using this endpoint we can find out information like what studies fit our search criteria, and also get data that we can filter to have only subjects from multiple studies, or only subjects from single studies.\n",
    "\n",
    "Any given subject will have one row per study they participated in. The subject_id in the last column of this view is the same as the `id` in the first column of the Subjects endpoint results. You can use this to combine information across endpoints, which is covered in the [Merging Results](../MergingResults.ipynb) notebook.\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a research study, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subject’s privacy. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this researchsubject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the researchsubject had there</li>\n",
    "  <li><b>member_of_research_project:</b> The name of the study/project that the subject particpated in</li>\n",
    "  <li><b>primary_diagnosis_condition:</b> The cancer, disease or other condition under study</li>\n",
    "  <li><b>primary_diagnosis_site:</b> The primary_disease_site that qualifies the researchsubject for the research_project</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47372646",
   "metadata": {},
   "source": [
    "### diagnosis\n",
    "\n",
    "The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f80fa21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3452 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>morphology</th>\n",
       "      <th>stage</th>\n",
       "      <th>grade</th>\n",
       "      <th>method_of_diagnosis</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>02ce9b1f-18f4-45ec-a487-63f9c74bb422</td>\n",
       "      <td>[{'system': 'GDC', 'value': '02ce9b1f-18f4-45e...</td>\n",
       "      <td>Mixed germ cell tumor</td>\n",
       "      <td>None</td>\n",
       "      <td>9085/3</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-MSK-P-0002849</td>\n",
       "      <td>3ecb27b2-75cc-44d0-879e-035e19f39193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0f7bb10c-a17e-4a98-b559-461edbf840c8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0f7bb10c-a17e-4a9...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>23121</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>C3N-01366</td>\n",
       "      <td>a1da2f64-5a24-4e6b-abc1-e70aae23082c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13c464b8-bd62-5d47-91ff-4b2c9183c48b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '13c464b8-bd62-5d4...</td>\n",
       "      <td>Oligodendroglioma, NOS</td>\n",
       "      <td>12764</td>\n",
       "      <td>9450/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-5870</td>\n",
       "      <td>eb42625e-fbbb-4bee-bbe9-315b3ca2138a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>195c5853-a955-551b-8601-6f0d8a08cea8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '195c5853-a955-551...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>14838</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-08-0351</td>\n",
       "      <td>91063e7e-60db-47a5-aeca-af7c37b1074e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2faa8b42-f38e-50f2-9920-cafeafc989ae</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2faa8b42-f38e-50f...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>28696</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-02-0281</td>\n",
       "      <td>42fbe95f-9068-45a8-a8a2-c24db757eae6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>7a3112c5-a878-5b23-bdbc-3aa91d2cede6</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7a3112c5-a878-5b2...</td>\n",
       "      <td>Mixed glioma</td>\n",
       "      <td>20128</td>\n",
       "      <td>9382/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-P5-A77X</td>\n",
       "      <td>878ced76-3144-4141-896f-50e3e863df36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>7b8b418c-70eb-5b2c-a880-1c869f324f21</td>\n",
       "      <td>[{'system': 'GDC', 'value': '7b8b418c-70eb-5b2...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>22332</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-06-1800</td>\n",
       "      <td>b1ce95bf-7441-4aac-8131-19a177d2b8ac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>8e43bc6d-48f0-5567-8333-64c91a50d4eb</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8e43bc6d-48f0-556...</td>\n",
       "      <td>Astrocytoma, anaplastic</td>\n",
       "      <td>16300</td>\n",
       "      <td>9401/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-P5-A736</td>\n",
       "      <td>ac3779f6-cc8b-428b-9531-03562a614515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>91404a49-c46a-43c3-8bdc-6ef7e4feba82</td>\n",
       "      <td>[{'system': 'GDC', 'value': '91404a49-c46a-43c...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>25423</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>C3N-01857</td>\n",
       "      <td>9b84c1a2-d24d-4085-b597-eff3d2203756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>b82d770d-caf9-5420-bcc1-3ab57dbdaf9c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b82d770d-caf9-542...</td>\n",
       "      <td>Mixed glioma</td>\n",
       "      <td>12853</td>\n",
       "      <td>9382/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A7QY</td>\n",
       "      <td>94a2ff09-f7ac-4aea-a2b2-0dc1b82ae635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   02ce9b1f-18f4-45ec-a487-63f9c74bb422   \n",
       "1   0f7bb10c-a17e-4a98-b559-461edbf840c8   \n",
       "2   13c464b8-bd62-5d47-91ff-4b2c9183c48b   \n",
       "3   195c5853-a955-551b-8601-6f0d8a08cea8   \n",
       "4   2faa8b42-f38e-50f2-9920-cafeafc989ae   \n",
       "..                                   ...   \n",
       "95  7a3112c5-a878-5b23-bdbc-3aa91d2cede6   \n",
       "96  7b8b418c-70eb-5b2c-a880-1c869f324f21   \n",
       "97  8e43bc6d-48f0-5567-8333-64c91a50d4eb   \n",
       "98  91404a49-c46a-43c3-8bdc-6ef7e4feba82   \n",
       "99  b82d770d-caf9-5420-bcc1-3ab57dbdaf9c   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '02ce9b1f-18f4-45e...   \n",
       "1   [{'system': 'GDC', 'value': '0f7bb10c-a17e-4a9...   \n",
       "2   [{'system': 'GDC', 'value': '13c464b8-bd62-5d4...   \n",
       "3   [{'system': 'GDC', 'value': '195c5853-a955-551...   \n",
       "4   [{'system': 'GDC', 'value': '2faa8b42-f38e-50f...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '7a3112c5-a878-5b2...   \n",
       "96  [{'system': 'GDC', 'value': '7b8b418c-70eb-5b2...   \n",
       "97  [{'system': 'GDC', 'value': '8e43bc6d-48f0-556...   \n",
       "98  [{'system': 'GDC', 'value': '91404a49-c46a-43c...   \n",
       "99  [{'system': 'GDC', 'value': 'b82d770d-caf9-542...   \n",
       "\n",
       "          primary_diagnosis age_at_diagnosis morphology stage         grade  \\\n",
       "0     Mixed germ cell tumor             None     9085/3  None  Not Reported   \n",
       "1              Glioblastoma            23121     9440/3  None  Not Reported   \n",
       "2    Oligodendroglioma, NOS            12764     9450/3  None  not reported   \n",
       "3              Glioblastoma            14838     9440/3  None  not reported   \n",
       "4              Glioblastoma            28696     9440/3  None  not reported   \n",
       "..                      ...              ...        ...   ...           ...   \n",
       "95             Mixed glioma            20128     9382/3  None  not reported   \n",
       "96             Glioblastoma            22332     9440/3  None  not reported   \n",
       "97  Astrocytoma, anaplastic            16300     9401/3  None  not reported   \n",
       "98             Glioblastoma            25423     9440/3  None  Not Reported   \n",
       "99             Mixed glioma            12853     9382/3  None  not reported   \n",
       "\n",
       "   method_of_diagnosis           subject_id  \\\n",
       "0                 None  GENIE-MSK-P-0002849   \n",
       "1                 None            C3N-01366   \n",
       "2                 None         TCGA-DU-5870   \n",
       "3                 None         TCGA-08-0351   \n",
       "4                 None         TCGA-02-0281   \n",
       "..                 ...                  ...   \n",
       "95                None         TCGA-P5-A77X   \n",
       "96                None         TCGA-06-1800   \n",
       "97                None         TCGA-P5-A736   \n",
       "98                None            C3N-01857   \n",
       "99                None         TCGA-S9-A7QY   \n",
       "\n",
       "                      researchsubject_id  \n",
       "0   3ecb27b2-75cc-44d0-879e-035e19f39193  \n",
       "1   a1da2f64-5a24-4e6b-abc1-e70aae23082c  \n",
       "2   eb42625e-fbbb-4bee-bbe9-315b3ca2138a  \n",
       "3   91063e7e-60db-47a5-aeca-af7c37b1074e  \n",
       "4   42fbe95f-9068-45a8-a8a2-c24db757eae6  \n",
       "..                                   ...  \n",
       "95  878ced76-3144-4141-896f-50e3e863df36  \n",
       "96  b1ce95bf-7441-4aac-8131-19a177d2b8ac  \n",
       "97  ac3779f6-cc8b-428b-9531-03562a614515  \n",
       "98  9b84c1a2-d24d-4085-b597-eff3d2203756  \n",
       "99  94a2ff09-f7ac-4aea-a2b2-0dc1b82ae635  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagnosisresults = myquery.diagnosis.run()\n",
    "diagnosisresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6008331",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Diagnosis Field Definitions</h3>\n",
    "\n",
    "<i>A diagnosis is a medical classification of a disease for a given research subject in a given study. A single research subject may have different diagnoses across different studies</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this diagnosis in this research subject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the diagnosed researchsubject had there</li>\n",
    "  <li><b>primary_diagnosis:</b> The main medical diagnosis for this subject in this study</li>\n",
    "  <li><b>age_at_diagnosis:</b> The subjects age in days after birth on the day they were first diagnosed</li>\n",
    "  <li><b>morphology:</b> The <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> diagnosic code for this diagnosis</li>\n",
    "  <li><b>stage:</b> A measure of disease spread. Different diseases may use different staging criteria, please refer to the originating data source to see what staging system is reported</li>\n",
    "  <li><b>grade:</b> A measure of cell abnormality. Different diseases may use different grading criteria, please refer to the originating data source to see what grading system is reported</li>\n",
    "  <li><b>method_of_diagnosis:</b> The test or system used for determining the diagnosis</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5855d4dd",
   "metadata": {},
   "source": [
    "### treatment\n",
    "\n",
    "The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d81bdba0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3535 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>days_to_treatment_start</th>\n",
       "      <th>days_to_treatment_end</th>\n",
       "      <th>therapeutic_agent</th>\n",
       "      <th>treatment_anatomic_site</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_end_reason</th>\n",
       "      <th>number_of_cycles</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>researchsubject_diagnosis_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0611032f-6cde-5d42-b305-23219b4b486a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0611032f-6cde-5d4...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-08-0346</td>\n",
       "      <td>5536201c-e739-46e6-8200-0bdcc28ae9ef</td>\n",
       "      <td>202921e5-a4e8-5d23-9d3c-de53394f93a9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>08cabbfd-c1c2-5689-a1e6-ae03f0e8ecc9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '08cabbfd-c1c2-568...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-RY-A845</td>\n",
       "      <td>36615695-921c-4c8a-93d6-6c788079beae</td>\n",
       "      <td>cbf0dbb0-8f9c-549d-b766-23bf00479cc7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1e45d588-c532-5952-b8e5-56cc67267853</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1e45d588-c532-595...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-8158</td>\n",
       "      <td>b8df1577-2a75-4eb7-ab27-f2dd76160730</td>\n",
       "      <td>7481b252-5efc-5ed9-9d2a-b049302a3869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2412b555-02be-5e02-8c26-cc4db13cb11c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2412b555-02be-5e0...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7477</td>\n",
       "      <td>38f4b9f3-6709-40ff-9a81-5fa2111b1ce6</td>\n",
       "      <td>f15ab57f-5079-5386-9a7b-9dfb1a8f2e1c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2afa81c1-efba-5fcd-b171-74820ac8d991</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2afa81c1-efba-5fc...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-32-2615</td>\n",
       "      <td>30011f30-2926-47f2-a8f3-4f1b58ffa227</td>\n",
       "      <td>c57d8c00-a082-56c0-9d5e-389441cc9523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>34b6e7f0-6cca-5054-90a0-fb65073ebca9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '34b6e7f0-6cca-505...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-WH-A86K</td>\n",
       "      <td>42156c6a-416e-4832-9c95-cee6084b0910</td>\n",
       "      <td>42519c28-1489-53d6-8dc6-907f0b377f71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3c9a6945-d2a1-526c-aab5-edd8ecf3152e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3c9a6945-d2a1-526...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HT-7879</td>\n",
       "      <td>c916b9a4-008e-4d79-a50c-f15c3549fcf6</td>\n",
       "      <td>616d6f27-c72a-59db-b7fd-a2f7962ee604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>4440d709-1beb-5233-9781-ae767e91b0de</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4440d709-1beb-523...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-12-1598</td>\n",
       "      <td>2b90648c-c99e-4d4d-8d9b-67ff2c4e6115</td>\n",
       "      <td>62f86760-df1c-5925-8b1a-e13d623a0183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>48f9d02d-5dfe-5124-85d9-ec9fdd7d8f77</td>\n",
       "      <td>[{'system': 'GDC', 'value': '48f9d02d-5dfe-512...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-06-1084</td>\n",
       "      <td>ac3582a9-46b1-4322-95fa-9e8b149b8044</td>\n",
       "      <td>d3175d4f-7edc-574e-8682-1a5a86427514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>4e7e5d91-f4fa-5ee1-89cf-45578413b59a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4e7e5d91-f4fa-5ee...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-E1-A7YU</td>\n",
       "      <td>fc3948fb-df27-4149-bbfb-04443dfecc19</td>\n",
       "      <td>2bf93f76-741f-5d39-81c6-17db692d53de</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0611032f-6cde-5d42-b305-23219b4b486a   \n",
       "1   08cabbfd-c1c2-5689-a1e6-ae03f0e8ecc9   \n",
       "2   1e45d588-c532-5952-b8e5-56cc67267853   \n",
       "3   2412b555-02be-5e02-8c26-cc4db13cb11c   \n",
       "4   2afa81c1-efba-5fcd-b171-74820ac8d991   \n",
       "..                                   ...   \n",
       "95  34b6e7f0-6cca-5054-90a0-fb65073ebca9   \n",
       "96  3c9a6945-d2a1-526c-aab5-edd8ecf3152e   \n",
       "97  4440d709-1beb-5233-9781-ae767e91b0de   \n",
       "98  48f9d02d-5dfe-5124-85d9-ec9fdd7d8f77   \n",
       "99  4e7e5d91-f4fa-5ee1-89cf-45578413b59a   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '0611032f-6cde-5d4...   \n",
       "1   [{'system': 'GDC', 'value': '08cabbfd-c1c2-568...   \n",
       "2   [{'system': 'GDC', 'value': '1e45d588-c532-595...   \n",
       "3   [{'system': 'GDC', 'value': '2412b555-02be-5e0...   \n",
       "4   [{'system': 'GDC', 'value': '2afa81c1-efba-5fc...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '34b6e7f0-6cca-505...   \n",
       "96  [{'system': 'GDC', 'value': '3c9a6945-d2a1-526...   \n",
       "97  [{'system': 'GDC', 'value': '4440d709-1beb-523...   \n",
       "98  [{'system': 'GDC', 'value': '48f9d02d-5dfe-512...   \n",
       "99  [{'system': 'GDC', 'value': '4e7e5d91-f4fa-5ee...   \n",
       "\n",
       "                 treatment_type treatment_outcome days_to_treatment_start  \\\n",
       "0        Radiation Therapy, NOS              None                    None   \n",
       "1   Pharmaceutical Therapy, NOS              None                    None   \n",
       "2   Pharmaceutical Therapy, NOS              None                    None   \n",
       "3   Pharmaceutical Therapy, NOS              None                    None   \n",
       "4        Radiation Therapy, NOS              None                    None   \n",
       "..                          ...               ...                     ...   \n",
       "95       Radiation Therapy, NOS              None                    None   \n",
       "96       Radiation Therapy, NOS              None                    None   \n",
       "97  Pharmaceutical Therapy, NOS              None                    None   \n",
       "98       Radiation Therapy, NOS              None                    None   \n",
       "99       Radiation Therapy, NOS              None                    None   \n",
       "\n",
       "   days_to_treatment_end therapeutic_agent treatment_anatomic_site  \\\n",
       "0                   None              None                    None   \n",
       "1                   None              None                    None   \n",
       "2                   None              None                    None   \n",
       "3                   None              None                    None   \n",
       "4                   None              None                    None   \n",
       "..                   ...               ...                     ...   \n",
       "95                  None              None                    None   \n",
       "96                  None              None                    None   \n",
       "97                  None              None                    None   \n",
       "98                  None              None                    None   \n",
       "99                  None              None                    None   \n",
       "\n",
       "   treatment_effect treatment_end_reason number_of_cycles    subject_id  \\\n",
       "0              None                 None             None  TCGA-08-0346   \n",
       "1              None                 None             None  TCGA-RY-A845   \n",
       "2              None                 None             None  TCGA-DU-8158   \n",
       "3              None                 None             None  TCGA-HT-7477   \n",
       "4              None                 None             None  TCGA-32-2615   \n",
       "..              ...                  ...              ...           ...   \n",
       "95             None                 None             None  TCGA-WH-A86K   \n",
       "96             None                 None             None  TCGA-HT-7879   \n",
       "97             None                 None             None  TCGA-12-1598   \n",
       "98             None                 None             None  TCGA-06-1084   \n",
       "99             None                 None             None  TCGA-E1-A7YU   \n",
       "\n",
       "                      researchsubject_id          researchsubject_diagnosis_id  \n",
       "0   5536201c-e739-46e6-8200-0bdcc28ae9ef  202921e5-a4e8-5d23-9d3c-de53394f93a9  \n",
       "1   36615695-921c-4c8a-93d6-6c788079beae  cbf0dbb0-8f9c-549d-b766-23bf00479cc7  \n",
       "2   b8df1577-2a75-4eb7-ab27-f2dd76160730  7481b252-5efc-5ed9-9d2a-b049302a3869  \n",
       "3   38f4b9f3-6709-40ff-9a81-5fa2111b1ce6  f15ab57f-5079-5386-9a7b-9dfb1a8f2e1c  \n",
       "4   30011f30-2926-47f2-a8f3-4f1b58ffa227  c57d8c00-a082-56c0-9d5e-389441cc9523  \n",
       "..                                   ...                                   ...  \n",
       "95  42156c6a-416e-4832-9c95-cee6084b0910  42519c28-1489-53d6-8dc6-907f0b377f71  \n",
       "96  c916b9a4-008e-4d79-a50c-f15c3549fcf6  616d6f27-c72a-59db-b7fd-a2f7962ee604  \n",
       "97  2b90648c-c99e-4d4d-8d9b-67ff2c4e6115  62f86760-df1c-5925-8b1a-e13d623a0183  \n",
       "98  ac3582a9-46b1-4322-95fa-9e8b149b8044  d3175d4f-7edc-574e-8682-1a5a86427514  \n",
       "99  fc3948fb-df27-4149-bbfb-04443dfecc19  2bf93f76-741f-5d39-81c6-17db692d53de  \n",
       "\n",
       "[100 rows x 14 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatmentresults = myquery.treatment.run()\n",
    "treatmentresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30804682",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Treatment Field Definitions</h3>\n",
    "\n",
    "<i><i>A treatment is a medical intervention for a diagnosed disease in a given subject in a given study. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this treatment of this diagnosis in this research subject</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the treated researchsubject had there</li>\n",
    "  <li><b>treatment_type:</b> The medical intervention undertaken</li>\n",
    "  <li><b>treatment_outcome:</b> The result of the medical intervention</li>\n",
    "  <li><b>days_to_treatment_start:</b> </li>\n",
    "  <li><b>days_to_treatment_end:</b> </li>\n",
    "  <li><b>therapeutic_agent:</b> What treatment or drug was used for this researchsubject</li>\n",
    "  <li><b>treatment_anatomic_site:</b> The specific body location of the treatment</li>\n",
    "  <li><b>treatment_effect:</b> </li>\n",
    "  <li><b>treatment_end_reason:</b> </li>\n",
    "  <li><b>number_of_cycles:</b> </li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the researchsubject. Can be joined to the `id` field from researchsubject results</li>\n",
    "  <li><b>researchsubject_diagnosis_id:</b> An identifier for the diagnosis. Can be joined to the `id` field from diagnosis results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c00bec8",
   "metadata": {},
   "source": [
    "### specimens\n",
    "\n",
    "We can use this same query to see what specimens are available for brain tissue at the CDA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51960eed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3511 ms\n",
      "\n",
      "            QueryID: e348a333-9d2a-419c-99b4-bb4fd44b7ce2\n",
      "            \n",
      "            Offset: 0\n",
      "            Count: 100\n",
      "            Total Row Count: 39150\n",
      "            More pages: True\n",
      "            \n"
     ]
    }
   ],
   "source": [
    "specresults =  myquery.specimen.run()\n",
    "print(specresults)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f254547",
   "metadata": {},
   "source": [
    "Nearly 40,000 specimens meet our search criteria! We would typically expect this number to be much larger than our number of subjects or researchsubjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. Since we didn't specify any further filters, our results will return all of these as seperate speciments. Let's look at a few:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c55b0128",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>primary_disease_type</th>\n",
       "      <th>anatomical_site</th>\n",
       "      <th>source_material_type</th>\n",
       "      <th>specimen_type</th>\n",
       "      <th>derived_from_specimen</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>018966a4-712b-4921-a7e7-02e390f43288</td>\n",
       "      <td>[{'system': 'GDC', 'value': '018966a4-712b-492...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-14729</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>analyte</td>\n",
       "      <td>790f03cf-ad8f-4cde-8ba5-f465837c687f</td>\n",
       "      <td>TCGA-06-0139</td>\n",
       "      <td>5bbeb8a7-3ac2-4ef1-bd48-6f83bc994a75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0190d8c0-20ab-5475-8804-c99979e77dbe</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0190d8c0-20ab-547...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>-14716</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>portion</td>\n",
       "      <td>5665435a-9b77-4504-a60a-b6d3edeb341f</td>\n",
       "      <td>TCGA-R8-A6MK</td>\n",
       "      <td>65836b59-b469-4b89-adf5-e65735e494f5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>01ed9fd7-7d44-401b-a6b7-4fc6922813fd</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01ed9fd7-7d44-401...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-26247</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>37af7534-d545-45ee-9de3-387ad85d98f7</td>\n",
       "      <td>TCGA-06-0644</td>\n",
       "      <td>652a39f0-c539-4303-bedc-f0d7f4669d9f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>02117459-65da-4a3c-ba9a-ead2c0daa5ba</td>\n",
       "      <td>[{'system': 'GDC', 'value': '02117459-65da-4a3...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>-17501</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>portion</td>\n",
       "      <td>fa252999-283a-4ddc-84b5-17257939cc73</td>\n",
       "      <td>TCGA-RY-A840</td>\n",
       "      <td>738143b3-2d82-42f0-9fd4-8cbabc9c9b4f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>023d5905-9dbf-40a7-90c1-761d3f48ad9a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '023d5905-9dbf-40a...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-22417</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>analyte</td>\n",
       "      <td>19496c06-447b-437a-bae8-20d1b0c411c5</td>\n",
       "      <td>TCGA-02-0046</td>\n",
       "      <td>5b8f05fa-0145-4142-8d41-5980e4d36d81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>485cb0f3-e668-47a5-8b8a-0f8f81688ef2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '485cb0f3-e668-47a...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-21768</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>6b524320-94c8-4035-b1d1-e1d3d3fe8679</td>\n",
       "      <td>TCGA-12-5301</td>\n",
       "      <td>10efeb3c-ef55-4227-91b9-8c2bb4ff43d5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>49280503-5136-4c10-8e42-061abbb6fa05</td>\n",
       "      <td>[{'system': 'GDC', 'value': '49280503-5136-4c1...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-27111</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>3b705faf-6584-4dd8-b5f5-423ec3ee7e60</td>\n",
       "      <td>TCGA-26-5134</td>\n",
       "      <td>20bad0d5-3135-49a4-a119-fa7a1e56fd1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>49289dc2-31a6-427d-a7b9-e8e3613e17c1</td>\n",
       "      <td>[{'system': 'GDC', 'value': '49289dc2-31a6-427...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>-16280</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>d1469517-114e-41da-9b50-ccecea62826b</td>\n",
       "      <td>TCGA-HT-7884</td>\n",
       "      <td>86aa5790-9afd-4653-926a-625516df1f56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>49ae4c42-847f-4da6-9a41-b9eaaa8831f1</td>\n",
       "      <td>[{'system': 'GDC', 'value': '49ae4c42-847f-4da...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>-17432</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>c55220e7-4667-42a2-a8de-4641a29cae1d</td>\n",
       "      <td>TCGA-EZ-7264</td>\n",
       "      <td>fc3b7596-f515-446f-81db-fed0154ca2c5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>4a02dbff-6707-4e83-9d79-efe01846f763</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4a02dbff-6707-4e8...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>-18195</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>6a6dc2fe-41ed-4d11-9ead-a2c1a1430f5b</td>\n",
       "      <td>TCGA-16-1045</td>\n",
       "      <td>cd73f38a-f6ff-4e7b-bd43-ec7b82dbe7df</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   018966a4-712b-4921-a7e7-02e390f43288   \n",
       "1   0190d8c0-20ab-5475-8804-c99979e77dbe   \n",
       "2   01ed9fd7-7d44-401b-a6b7-4fc6922813fd   \n",
       "3   02117459-65da-4a3c-ba9a-ead2c0daa5ba   \n",
       "4   023d5905-9dbf-40a7-90c1-761d3f48ad9a   \n",
       "..                                   ...   \n",
       "95  485cb0f3-e668-47a5-8b8a-0f8f81688ef2   \n",
       "96  49280503-5136-4c10-8e42-061abbb6fa05   \n",
       "97  49289dc2-31a6-427d-a7b9-e8e3613e17c1   \n",
       "98  49ae4c42-847f-4da6-9a41-b9eaaa8831f1   \n",
       "99  4a02dbff-6707-4e83-9d79-efe01846f763   \n",
       "\n",
       "                                           identifier associated_project  \\\n",
       "0   [{'system': 'GDC', 'value': '018966a4-712b-492...           TCGA-GBM   \n",
       "1   [{'system': 'GDC', 'value': '0190d8c0-20ab-547...           TCGA-LGG   \n",
       "2   [{'system': 'GDC', 'value': '01ed9fd7-7d44-401...           TCGA-GBM   \n",
       "3   [{'system': 'GDC', 'value': '02117459-65da-4a3...           TCGA-LGG   \n",
       "4   [{'system': 'GDC', 'value': '023d5905-9dbf-40a...           TCGA-GBM   \n",
       "..                                                ...                ...   \n",
       "95  [{'system': 'GDC', 'value': '485cb0f3-e668-47a...           TCGA-GBM   \n",
       "96  [{'system': 'GDC', 'value': '49280503-5136-4c1...           TCGA-GBM   \n",
       "97  [{'system': 'GDC', 'value': '49289dc2-31a6-427...           TCGA-LGG   \n",
       "98  [{'system': 'GDC', 'value': '49ae4c42-847f-4da...           TCGA-LGG   \n",
       "99  [{'system': 'GDC', 'value': '4a02dbff-6707-4e8...           TCGA-GBM   \n",
       "\n",
       "   age_at_collection primary_disease_type anatomical_site  \\\n",
       "0             -14729              Gliomas            None   \n",
       "1             -14716              Gliomas            None   \n",
       "2             -26247              Gliomas            None   \n",
       "3             -17501              Gliomas            None   \n",
       "4             -22417              Gliomas            None   \n",
       "..               ...                  ...             ...   \n",
       "95            -21768              Gliomas            None   \n",
       "96            -27111              Gliomas            None   \n",
       "97            -16280              Gliomas            None   \n",
       "98            -17432              Gliomas            None   \n",
       "99            -18195              Gliomas            None   \n",
       "\n",
       "    source_material_type specimen_type                 derived_from_specimen  \\\n",
       "0   Blood Derived Normal       analyte  790f03cf-ad8f-4cde-8ba5-f465837c687f   \n",
       "1          Primary Tumor       portion  5665435a-9b77-4504-a60a-b6d3edeb341f   \n",
       "2          Primary Tumor         slide  37af7534-d545-45ee-9de3-387ad85d98f7   \n",
       "3   Blood Derived Normal       portion  fa252999-283a-4ddc-84b5-17257939cc73   \n",
       "4   Blood Derived Normal       analyte  19496c06-447b-437a-bae8-20d1b0c411c5   \n",
       "..                   ...           ...                                   ...   \n",
       "95         Primary Tumor       aliquot  6b524320-94c8-4035-b1d1-e1d3d3fe8679   \n",
       "96         Primary Tumor       aliquot  3b705faf-6584-4dd8-b5f5-423ec3ee7e60   \n",
       "97         Primary Tumor       aliquot  d1469517-114e-41da-9b50-ccecea62826b   \n",
       "98  Blood Derived Normal       aliquot  c55220e7-4667-42a2-a8de-4641a29cae1d   \n",
       "99         Primary Tumor       aliquot  6a6dc2fe-41ed-4d11-9ead-a2c1a1430f5b   \n",
       "\n",
       "      subject_id                    researchsubject_id  \n",
       "0   TCGA-06-0139  5bbeb8a7-3ac2-4ef1-bd48-6f83bc994a75  \n",
       "1   TCGA-R8-A6MK  65836b59-b469-4b89-adf5-e65735e494f5  \n",
       "2   TCGA-06-0644  652a39f0-c539-4303-bedc-f0d7f4669d9f  \n",
       "3   TCGA-RY-A840  738143b3-2d82-42f0-9fd4-8cbabc9c9b4f  \n",
       "4   TCGA-02-0046  5b8f05fa-0145-4142-8d41-5980e4d36d81  \n",
       "..           ...                                   ...  \n",
       "95  TCGA-12-5301  10efeb3c-ef55-4227-91b9-8c2bb4ff43d5  \n",
       "96  TCGA-26-5134  20bad0d5-3135-49a4-a119-fa7a1e56fd1b  \n",
       "97  TCGA-HT-7884  86aa5790-9afd-4653-926a-625516df1f56  \n",
       "98  TCGA-EZ-7264  fc3b7596-f515-446f-81db-fed0154ca2c5  \n",
       "99  TCGA-16-1045  cd73f38a-f6ff-4e7b-bd43-ec7b82dbe7df  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccb0f1c",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Specimen Field Definitions</h3>\n",
    "\n",
    "<i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this specimen</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the specimen had there</li>\n",
    "  <li><b>associated_project:</b> The name of the study/project that the subject particpated in</li>\n",
    "  <li><b>age_at_collection:</b> The subjects age in days (counting backwards to birth) on the day of the collection of the proximate specimen</li>\n",
    "  <li><b>primary_disease_type:</b> The disease that qualifies the researchsubject for the associated_project</li>\n",
    "  <li><b>anatomical_site:</b> The body part from which the proximate specimen was taken</li>\n",
    "  <li><b>source_material_type:</b> The general kind of material from which the specimen was derived, indicating the physical nature of the source material</li>\n",
    "  <li><b>specimen_type:</b> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide</li>\n",
    "  <li><b>derived_from_specimen:</b> For derived samples, the `id` for the original sample</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n",
    "### file\n",
    "\n",
    "The file endpoint returns information about files that meet our search criteria, regardless of whether they are attached to subjects, research-subjects or specimens: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "94a0361f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3522 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 1aa12ab0-ffe0-4588-8514-51c3960d8d5c\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 4530800\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.file.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6006ec89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3805 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1b450a73-a16d-4de4-b140-9295335063ad</td>\n",
       "      <td>[{'system': 'PDC', 'value': '1b450a73-a16d-4de...</td>\n",
       "      <td>NCI-11Plex-2-F12-f10321.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:1b450a73-a16d-4de4-b140-92953350...</td>\n",
       "      <td>2850423</td>\n",
       "      <td>5e82ca67c560b6ab0b175ee5057ca02e</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C952635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ef900de0-892c-4e2e-8eae-74caf55de54c</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ef900de0-892c-4e2...</td>\n",
       "      <td>TCGA-LGG.c6bdb34c-fa93-4736-a842-afec0d9fa990....</td>\n",
       "      <td>Structural Variation</td>\n",
       "      <td>Transcript Fusion</td>\n",
       "      <td>TSV</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:ef900de0-892c-4e2e-8eae-74caf55d...</td>\n",
       "      <td>15444</td>\n",
       "      <td>293a480b7106cc7945f5807f9eb961de</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-DH-5140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d43ef537-6cd0-446a-9bf0-fe0767ed6bcf</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd43ef537-6cd0-446...</td>\n",
       "      <td>TCGA-R8-A6YH-01A-02-TS2.AD2BC93E-924F-46FC-A41...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:d43ef537-6cd0-446a-9bf0-fe0767ed...</td>\n",
       "      <td>356728009</td>\n",
       "      <td>78c73f88728bf3c73fb148105eb9348f</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-R8-A6YH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>58581feb-03c1-496c-9328-e8dda42c0d59</td>\n",
       "      <td>[{'system': 'GDC', 'value': '58581feb-03c1-496...</td>\n",
       "      <td>5c6ea055-2352-4111-9f8d-6eec3b9b3847.methylati...</td>\n",
       "      <td>DNA Methylation</td>\n",
       "      <td>Methylation Beta Value</td>\n",
       "      <td>TXT</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>drs://dg.4DFC:58581feb-03c1-496c-9328-e8dda42c...</td>\n",
       "      <td>22617382</td>\n",
       "      <td>617b0a0fd0e824959643edd0a2fc75ff</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>phs001287</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-01334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1fd6817e-fa11-4569-a740-86584eaff394</td>\n",
       "      <td>[{'system': 'PDC', 'value': '1fd6817e-fa11-456...</td>\n",
       "      <td>10CPTAC_GBM_W_PNNL_20190501_B3S2_f09.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:1fd6817e-fa11-4569-a740-86584eaf...</td>\n",
       "      <td>4543929</td>\n",
       "      <td>11ab9596db36bd898520de88b4d7ad84</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-01368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>70b44166-c646-43eb-8d76-b567bf9328ac</td>\n",
       "      <td>[{'system': 'PDC', 'value': '70b44166-c646-43e...</td>\n",
       "      <td>NCI-11pex-6C-Phos-c05377.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:70b44166-c646-43eb-8d76-b567bf93...</td>\n",
       "      <td>220183305</td>\n",
       "      <td>756391f44696965c00f4902a5bdcf2ea</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C311067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>da7f4908-e98d-4ce6-93e3-a721276f733b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'da7f4908-e98d-4ce...</td>\n",
       "      <td>179f277a-8e72-4008-8a27-ea6e0d4ec7db.wxs.aliqu...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Aggregated Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>drs://dg.4DFC:da7f4908-e98d-4ce6-93e3-a721276f...</td>\n",
       "      <td>965606</td>\n",
       "      <td>e18a3fce6c7fc6bdc2d09858cc17bd1e</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-19-1787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>8341914b-205d-4765-91ad-3e223c465459</td>\n",
       "      <td>[{'system': 'PDC', 'value': '8341914b-205d-476...</td>\n",
       "      <td>10CPTAC_GBM_W_PNNL_20190501_B3S2_f24.mzid.gz</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzIdentML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:8341914b-205d-4765-91ad-3e223c46...</td>\n",
       "      <td>8766142</td>\n",
       "      <td>af764efe01d6ff020fcd8a9af690aed7</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-03184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ade18349-cd37-4da0-94c2-2a9e664002ac</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ade18349-cd37-4da...</td>\n",
       "      <td>962fbeae-2f42-4969-b35d-b1ec852c57ca.wxs.aliqu...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Masked Somatic Mutation</td>\n",
       "      <td>MAF</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>drs://dg.4DFC:ade18349-cd37-4da0-94c2-2a9e6640...</td>\n",
       "      <td>23065</td>\n",
       "      <td>ed693ebc68c8af3004ca47a36a1b4f4d</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-32-4211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>e0c450dd-04f1-48e9-b572-19606de73d5b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e0c450dd-04f1-48e...</td>\n",
       "      <td>6998fda2-aab1-4175-8606-2df81e5e7114_wxs_gdc_r...</td>\n",
       "      <td>Sequencing Reads</td>\n",
       "      <td>Aligned Reads</td>\n",
       "      <td>BAM</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>drs://dg.4DFC:e0c450dd-04f1-48e9-b572-19606de7...</td>\n",
       "      <td>16053004781</td>\n",
       "      <td>f7e90abfe1723bbdac0cd26ae790e579</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TCGA-DB-A4XE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1b450a73-a16d-4de4-b140-9295335063ad   \n",
       "1   ef900de0-892c-4e2e-8eae-74caf55de54c   \n",
       "2   d43ef537-6cd0-446a-9bf0-fe0767ed6bcf   \n",
       "3   58581feb-03c1-496c-9328-e8dda42c0d59   \n",
       "4   1fd6817e-fa11-4569-a740-86584eaff394   \n",
       "..                                   ...   \n",
       "95  70b44166-c646-43eb-8d76-b567bf9328ac   \n",
       "96  da7f4908-e98d-4ce6-93e3-a721276f733b   \n",
       "97  8341914b-205d-4765-91ad-3e223c465459   \n",
       "98  ade18349-cd37-4da0-94c2-2a9e664002ac   \n",
       "99  e0c450dd-04f1-48e9-b572-19606de73d5b   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'PDC', 'value': '1b450a73-a16d-4de...   \n",
       "1   [{'system': 'GDC', 'value': 'ef900de0-892c-4e2...   \n",
       "2   [{'system': 'GDC', 'value': 'd43ef537-6cd0-446...   \n",
       "3   [{'system': 'GDC', 'value': '58581feb-03c1-496...   \n",
       "4   [{'system': 'PDC', 'value': '1fd6817e-fa11-456...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'PDC', 'value': '70b44166-c646-43e...   \n",
       "96  [{'system': 'GDC', 'value': 'da7f4908-e98d-4ce...   \n",
       "97  [{'system': 'PDC', 'value': '8341914b-205d-476...   \n",
       "98  [{'system': 'GDC', 'value': 'ade18349-cd37-4da...   \n",
       "99  [{'system': 'GDC', 'value': 'e0c450dd-04f1-48e...   \n",
       "\n",
       "                                                label  \\\n",
       "0                         NCI-11Plex-2-F12-f10321.psm   \n",
       "1   TCGA-LGG.c6bdb34c-fa93-4736-a842-afec0d9fa990....   \n",
       "2   TCGA-R8-A6YH-01A-02-TS2.AD2BC93E-924F-46FC-A41...   \n",
       "3   5c6ea055-2352-4111-9f8d-6eec3b9b3847.methylati...   \n",
       "4            10CPTAC_GBM_W_PNNL_20190501_B3S2_f09.psm   \n",
       "..                                                ...   \n",
       "95                   NCI-11pex-6C-Phos-c05377.mzML.gz   \n",
       "96  179f277a-8e72-4008-8a27-ea6e0d4ec7db.wxs.aliqu...   \n",
       "97       10CPTAC_GBM_W_PNNL_20190501_B3S2_f24.mzid.gz   \n",
       "98  962fbeae-2f42-4969-b35d-b1ec852c57ca.wxs.aliqu...   \n",
       "99  6998fda2-aab1-4175-8606-2df81e5e7114_wxs_gdc_r...   \n",
       "\n",
       "                  data_category                    data_type file_format  \\\n",
       "0      Peptide Spectral Matches                         Text         tsv   \n",
       "1          Structural Variation            Transcript Fusion         TSV   \n",
       "2                   Biospecimen                  Slide Image         SVS   \n",
       "3               DNA Methylation       Methylation Beta Value         TXT   \n",
       "4      Peptide Spectral Matches                         Text         tsv   \n",
       "..                          ...                          ...         ...   \n",
       "95       Processed Mass Spectra                Open Standard        mzML   \n",
       "96  Simple Nucleotide Variation  Aggregated Somatic Mutation         MAF   \n",
       "97     Peptide Spectral Matches                Open Standard   mzIdentML   \n",
       "98  Simple Nucleotide Variation      Masked Somatic Mutation         MAF   \n",
       "99             Sequencing Reads                Aligned Reads         BAM   \n",
       "\n",
       "                                   associated_project  \\\n",
       "0   Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "1                                            TCGA-LGG   \n",
       "2                                            TCGA-LGG   \n",
       "3                                             CPTAC-3   \n",
       "4                                    CPTAC3-Discovery   \n",
       "..                                                ...   \n",
       "95  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "96                                           TCGA-GBM   \n",
       "97                                   CPTAC3-Discovery   \n",
       "98                                           TCGA-GBM   \n",
       "99                                           TCGA-LGG   \n",
       "\n",
       "                                              drs_uri    byte_size  \\\n",
       "0   drs://dg.4DFC:1b450a73-a16d-4de4-b140-92953350...      2850423   \n",
       "1   drs://dg.4DFC:ef900de0-892c-4e2e-8eae-74caf55d...        15444   \n",
       "2   drs://dg.4DFC:d43ef537-6cd0-446a-9bf0-fe0767ed...    356728009   \n",
       "3   drs://dg.4DFC:58581feb-03c1-496c-9328-e8dda42c...     22617382   \n",
       "4   drs://dg.4DFC:1fd6817e-fa11-4569-a740-86584eaf...      4543929   \n",
       "..                                                ...          ...   \n",
       "95  drs://dg.4DFC:70b44166-c646-43eb-8d76-b567bf93...    220183305   \n",
       "96  drs://dg.4DFC:da7f4908-e98d-4ce6-93e3-a721276f...       965606   \n",
       "97  drs://dg.4DFC:8341914b-205d-4765-91ad-3e223c46...      8766142   \n",
       "98  drs://dg.4DFC:ade18349-cd37-4da0-94c2-2a9e6640...        23065   \n",
       "99  drs://dg.4DFC:e0c450dd-04f1-48e9-b572-19606de7...  16053004781   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   5e82ca67c560b6ab0b175ee5057ca02e     Proteomic             None   \n",
       "1   293a480b7106cc7945f5807f9eb961de       Genomic             None   \n",
       "2   78c73f88728bf3c73fb148105eb9348f       Genomic             None   \n",
       "3   617b0a0fd0e824959643edd0a2fc75ff       Genomic             None   \n",
       "4   11ab9596db36bd898520de88b4d7ad84     Proteomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  756391f44696965c00f4902a5bdcf2ea     Proteomic             None   \n",
       "96  e18a3fce6c7fc6bdc2d09858cc17bd1e       Genomic             None   \n",
       "97  af764efe01d6ff020fcd8a9af690aed7     Proteomic             None   \n",
       "98  ed693ebc68c8af3004ca47a36a1b4f4d       Genomic             None   \n",
       "99  f7e90abfe1723bbdac0cd26ae790e579       Genomic             None   \n",
       "\n",
       "   dbgap_accession_number researchsubject_specimen_id researchsubject_id  \\\n",
       "0                    None                                                  \n",
       "1                    None                                                  \n",
       "2                    None                                                  \n",
       "3               phs001287                                                  \n",
       "4                    None                                                  \n",
       "..                    ...                         ...                ...   \n",
       "95                   None                                                  \n",
       "96                   None                                                  \n",
       "97                   None                                                  \n",
       "98                   None                                                  \n",
       "99                   None                                                  \n",
       "\n",
       "      subject_id  \n",
       "0        C952635  \n",
       "1   TCGA-DH-5140  \n",
       "2   TCGA-R8-A6YH  \n",
       "3      C3N-01334  \n",
       "4      C3N-01368  \n",
       "..           ...  \n",
       "95       C311067  \n",
       "96  TCGA-19-1787  \n",
       "97     C3N-03184  \n",
       "98  TCGA-32-4211  \n",
       "99  TCGA-DB-A4XE  \n",
       "\n",
       "[100 rows x 16 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileresults = myquery.file.run()\n",
    "fileresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46fcb3e4",
   "metadata": {},
   "source": [
    "As you might expect, searching file gives us a huge number of results. This is great if you are surveying what kind of data is available, but is less useful for getting a coherent cohort. \n",
    "\n",
    "A better way to get files for a specific cohort is to chain your queries together, which we cover in the next tutorial [Chaining Queries](../ChainingQueries): Combine information from multiple endpoints, and build And/Or/Like and other advanced query strings.\n",
    "\n",
    "Another useful way to look at high level information is to use our counts feature which returns summary information rather than the full search results. Check out the [Data Summaries tutorial](../DataSummaries) to try it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "549257b1",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div style=\"background-color:#a2f2ed;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The unique identifier for this file</li>\n",
    "  <li><b>identifier:</b> An embedded array of information that includes the originating data center and the ID the file had there</li>\n",
    "  <li><b>label:</b> The full name of the file</li>\n",
    "  <li><b>data_catagory:</b> A desecription of the kind of general kind data the file holds</li>\n",
    "  <li><b>data_type:</b> A more specific descripton of the data type</li>\n",
    "  <li><b>file_format:</b> String to identify the full file extension including compression extensions</li>\n",
    "  <li><b>associated_project:</b> The name the data center uses for the study this file was generated for</li>\n",
    "  <li><b>drs_uri:</b> A unique identifier that can be used to retreive this specific file from a server</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file</li>\n",
    "  <li><b>data_modality:</b> Describes the biological nature of the information gathered as the result of an activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\"</li>\n",
    "  <li><b>imaging_modality:</b> For files with the `data_modality` of \"Imaging\", a descriptor for the image type</li>\n",
    "  <li><b>dbgap_accession_number:</b> The project id number for this data on dbGaP</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
